packages/protocol/contracts/signal/SignalService.sol

```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: MIT
+//
+// ╭━━━━╮╱╱╭╮╱╱╱╱╱╭╮╱╱╱╱╱╭╮
+// ┃╭╮╭╮┃╱╱┃┃╱╱╱╱╱┃┃╱╱╱╱╱┃┃
+// ╰╯┃┃┣┻━┳┫┃╭┳━━╮┃┃╱╱╭━━┫╰━┳━━╮
+// ╱╱┃┃┃╭╮┣┫╰╯┫╭╮┃┃┃╱╭┫╭╮┃╭╮┃━━┫
+// ╱╱┃┃┃╭╮┃┃╭╮┫╰╯┃┃╰━╯┃╭╮┃╰╯┣━━┃
+// ╱╱╰╯╰╯╰┻┻╯╰┻━━╯╰━━━┻╯╰┻━━┻━━╯
+pragma solidity ^0.8.9;
+
+import "../common/EssentialContract.sol";
+import "../common/IHeaderSync.sol";
+import "../libs/LibBlockHeader.sol";
+import "../libs/LibTrieProof.sol";
+import "./ISignalService.sol";
+
+contract SignalService is ISignalService, EssentialContract {
+    using LibBlockHeader for BlockHeader;
+
+    struct SignalProof {
+        BlockHeader header;
+        bytes proof;
+    }
+
+    /// @dev Initializer to be called after being deployed behind a proxy.
+    function init(address _addressManager) external initializer {
+        EssentialContract._init(_addressManager);
+    }
+
+    function sendSignal(bytes32 signal) public returns (bytes32 storageSlot) {
+        require(signal != 0, "B:signal");
+
+        storageSlot = getSignalSlot(msg.sender, signal);
+        assembly {
+            sstore(storageSlot, 1)
+        }
+    }
+
+    function isSignalSent(
+        address app,
+        bytes32 signal
+    ) public view returns (bool) {
+        require(app != address(0), "B:app");
+        require(signal != 0, "B:signal");
+
+        bytes32 k = getSignalSlot(app,  signal);
+        uint256 v;
+        assembly {
+            v := sload(k)
+        }
+        return v == 1;
+    }
+
+    function isSignalReceived(
+        address app,
+        bytes32 signal,
+        bytes calldata proof
+    ) public view returns (bool received) {
+        require(app != address(0), "B:app");
+        require(signal != 0, "B:signal");
+
+        SignalProof memory sp = abi.decode(proof, (SignalProof));
+        received = LibTrieProof.verify({
+            stateRoot: sp.header.stateRoot,
+            addr: app,
```

Comment:
here we should use `srcChainSignalServiceAddress` instead of `app`? Since the slot was written in the src chain's `signalService`